crates/qpy/src/circuit_writer.rs

```diff
@@ -0,0 +1,910 @@
+    recognize_custom_operation, serialize_metadata,
+};
+use crate::value::{
+    GenericValue, QPYWriteData, bit_types, circuit_instruction_types, expression_var_declaration,
+    get_circuit_type_key, pack_generic_value, pack_standalone_var, pack_stretch, register_types,
+    serialize,
+};
+
+use crate::UnsupportedFeatureForVersion;
+
+/// packing the qubits and clbits of a specific instruction into CircuitInstructionArgPack
+pub fn get_packed_bit_list(
+    inst: &PackedInstruction,
+    circuit_data: &CircuitData,
+) -> Vec<formats::CircuitInstructionArgPack> {
+    let mut result: Vec<formats::CircuitInstructionArgPack> = Vec::new();
+    for qubit in circuit_data.get_qargs(inst.qubits).iter() {
+        result.push(formats::CircuitInstructionArgPack {
+            bit_type: bit_types::QUBIT,
+            index: (qubit.index() as u32),
+        });
+    }
+    for clbit in circuit_data.get_cargs(inst.clbits).iter() {
+        result.push(formats::CircuitInstructionArgPack {
+            bit_type: bit_types::CLBIT,
+            index: (clbit.index() as u32),
+        });
+    }
+    result
+}
+
+pub fn get_condition_data(
+    op: &PackedOperation,
+    qpy_data: &QPYWriteData,
+) -> PyResult<formats::ConditionPack> {
+    match op.view() {
+        OperationRef::Instruction(py_inst) => {
+            py_get_condition_data_from_inst(&py_inst.instruction, qpy_data)
+        }
+        // we assume only PyInstructions have condition data at this stage
+        _ => Ok(formats::ConditionPack {
+            key: formats::condition_types::NONE,
+            register_size: 0u16,
+            value: 0i64,
+            data: formats::ConditionData::None,
+        }),
+    }
+}
+
+/// pack all the instructions in the circuit, returning both the packed instructions
+/// and the dictionary of custom operations generated in the process
+pub fn pack_instructions(
+    qpy_data: &mut QPYWriteData,
+) -> PyResult<(
+    Vec<formats::CircuitInstructionV2Pack>,
+    HashMap<String, PackedOperation>,
+)> {
+    let mut custom_operations = HashMap::new();
+    let mut custom_new_operations = Vec::new();
+    let instructions = qpy_data.circuit_data.data().to_vec();
+    Ok((
+        instructions
+            .iter()
+            .map(|instruction| {
+                pack_instruction(
+                    instruction,
+                    &mut custom_operations,
+                    &mut custom_new_operations,
+                    qpy_data,
+                )
+            })
+            .collect::<PyResult<_>>()?,
+        custom_operations,
+    ))
+}
+
+/// packs one specific instruction into CircuitInstructionV2Pack, creating a new custom operation if needed
+pub fn pack_instruction(
+    instruction: &PackedInstruction,
+    custom_operations: &mut HashMap<String, PackedOperation>,
+    new_custom_operations: &mut Vec<String>,
+    qpy_data: &mut QPYWriteData,
+) -> PyResult<formats::CircuitInstructionV2Pack> {
```

Comment:
I think we should rather have a big match block over `instruction.op.view()` and handle getting info like class name, number of control qubits etc per match arm.  In simple cases (like standard gates) most of the info can be extracted directly, other cases will require helper functions but those would be simpler. For example, `get_condition_data` and `py_get_instruction_annotations` can just get `OperationRef::Instruction`, `gate_class_name` would be simpler etc..